### PR TITLE
Implement has_feature.

### PR DESCRIPTION
**Pull request by sunfishcode**
This implements `has_feature` as described in [the design](https://github.com/WebAssembly/design/blob/master/AstSemantics.md#feature-test).

### CODE REVIEW

**Comment by sunfishcode** [comment]
This is now rebased on top of the host opcode grouping changes, which required teaching the host opcode code about host opcodes that don't need memory.

**Comment by kg** [comment]
It would be cool to add a known-always-present feature so we can test both the positive and negative case in the automated test. That's a pretty small thing though. Otherwise this LGTM.

**Comment by sunfishcode** [comment]
Good idea. I added a feature called "wasm" which is claimed to be unconditionally supported :-).

File: ml-proto/spec/check.ml [hunk 1]
```diff
@@ -90,10 +90,17 @@ let type_cvt at = function
     | DemoteFloat64 -> error at "invalid conversion"
     ), Float64Type
 
+type type_hostop_type = {
```
Comment (rossberg-chromium) [inline ml-proto/spec/check.ml]: Instead of introducing another variant of function type for a singular use, can't `type_hostop` just return a pair `func_type * bool`?

File: ml-proto/spec/eval.ml [hunk 2]
```diff
@@ -298,6 +304,9 @@ and eval_hostop host mem hostop vs at =
     Memory.grow mem delta;
     None;
 
+  | HasFeature str, [] ->
+    Some (Int32 (if (host.has_feature str) then 1l else 0l))
```
Comment (rossberg-chromium) [inline ml-proto/spec/eval.ml]: Nit: redundant parens

**Comment by rossberg-chromium** [comment]
I'm afraid I still don't understand how a feature test _inside_ the language is supposed to make sense -- AFAICS it is utterly useless as long as there's no defined way to actually handle unsupported features _inside_ the language as well. And I have no idea how that would be done or specced or implemented. The design docs don't explain that either, only how to do feature testing on the meta level (which doesn't require the operator). It seems to me that putting in the operator now is starting with a (useless) step 2 before we have a clue about step 1.

But other than these fundamental doubts about the premise, and the nits above, LGTM. :)

**Comment by lukewagner** [comment]
@rossberg-chromium The intended usage I'm anticipating in the MVP is described [here](https://github.com/WebAssembly/design/blob/master/FutureFeatures.md#better-feature-testing-support).

**Comment by rossberg-chromium** [comment]
@lukewagner, yes, that's what I looked at, but as far as the MVP is concerned, the only workable suggestions there are doing meta-level feature detection (or am I misreading that?). So what still puzzles me is how _in-language_ feature detection is useful before you also add the actual ability to "decode-through" unrecognised features?

**Comment by lukewagner** [comment]
@rossberg-chromium Because we'd want implement layer 1 decoder (which also performed the polyfilling) in wasm.  That (tiny) module needs to learn what features are supported *somehow*, so might as well be from within wasm; otherwise we'd have to add it as a JS feature and that seems weird.

**Comment by rossberg-chromium** [comment]
I see! Okay, sorry that I didn't get that from the description. I wouldn't mind clarifying this use case (and perhaps moving the explanation to the FeatureTest page itself). ;)

**Comment by lukewagner** [comment]
@rossberg-chromium Will do; I forgot we even had FeatureTest.md :)

**Comment by kg** [comment]
> So what still puzzles me is how in-language feature detection is useful before you also add the actual ability to "decode-through" unrecognised features?

For the record, my binary encoding proposal already covers being able to decode unrecognized AST nodes.

File: ml-proto/spec/check.ml (hunk 1 above)
Comment (sunfishcode) [inline ml-proto/spec/check.ml]: That works.

**Comment by sunfishcode** [comment]
The patch is updated to address the comments above. Any further comments?

**Comment by rossberg-chromium** [comment]
lgtm
